[java] Fix JUnitSpellingRule false positive

KroArtem opened the pull request:
## Describe the PR

    Do not trigger if it's a junit5 class (isJUnit5Class from AbstractJUnitRule can be used).
    Do not trigger if junit4 doesn't extend TestCase

It'd be probably good to add more explanation to the rule itself.
<!-- A clear and concise description of the bug the PR fixes or the feature the PR introduces. -->

## Related issues

<!-- PR relates to issues in the `pmd` repo: -->

- Fixes #2477

## Ready?

<!-- If you feel like you can help to check off the following tasks, that'd be great. If not, don't worry - we will take care of it. -->

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by travis)
- [ ] Added (in-code) documentation (if needed)

pmd-test commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This changeset introduces 0 ne...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 286 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/54d513afcb8640a68403f1eb5a880ac6/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

adangel commented on pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/AbstractJUnitRule.java:
```diff
@@ -89,7 +89,7 @@ private boolean isJUnit3Method(ASTMethodDeclaration method) {
     }
 
-    private boolean isJUnit3Class(ASTCompilationUnit node) {
+    protected boolean isJUnit3Class(ASTCompilationUnit node) {
```
I'd not expose these methods.
These methods are called already when visiting CompilationUnit and the result is stored in fields, see lines 38-40 above.
I'd make these fields protected, so that you can use it in the rule.

KroArtem replied:
Indeed, fixed.

adangel commented on pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/JUnitSpellingRule.java:
```diff
@@ -5,13 +5,19 @@
 
     @Override
     public Object visit(ASTMethodDeclaration node, Object data) {
+        ASTCompilationUnit acu = node.getFirstParentOfType(ASTCompilationUnit.class);
+        if (isJUnit5Class(acu) || isJUnit4Class(acu)) {
```
If using the protected fields from the base class, you don't need to search here again for the compilation unit and call again the methods - that was already done....

KroArtem replied:
You're right

KroArtem commented:
@adangel , I've changed the description to this one:
```
<description>
            In JUnit 3, the setUp method is used to set up all data entities required in running tests.
            The tearDown method is used to clean up all data entities required in running tests.
            You should not misspell method name if you want your test to set up and clean up everything correctly.
</description>
```
Other remarks are also fixed.
